firecrown/likelihood/gauss_family/gauss_family.py

```diff
@@ -98,28 +115,50 @@ def read(self, sacc_data: sacc.Sacc) -> None:
             raise RuntimeError(msg)
 
         covariance = sacc_data.covariance.dense
+
+        indices_list = []
+        data_vector_list = []
         for stat in self.statistics:
             stat.read(sacc_data)
+            if stat.statistic.sacc_indices is None:
+                raise RuntimeError(
+                    f"The statistic {stat.statistic} has no sacc_indices."
+                )
+            indices_list.append(stat.statistic.sacc_indices.copy())
+            data_vector_list.append(stat.statistic.get_data_vector())
 
-        indices_list = [s.statistic.sacc_indices.copy() for s in self.statistics]
         indices = np.concatenate(indices_list)
+        data_vector = np.concatenate(data_vector_list)
         cov = np.zeros((len(indices), len(indices)))
 
         for new_i, old_i in enumerate(indices):
             for new_j, old_j in enumerate(indices):
                 cov[new_i, new_j] = covariance[old_i, old_j]
 
+        self.data_vector = data_vector
+        self.cov_index_map = {old_i: new_i for new_i, old_i in enumerate(indices)}
         self.cov = cov
         self.cholesky = scipy.linalg.cholesky(self.cov, lower=True)
         self.inv_cov = np.linalg.inv(cov)
 
         self.state = State.READY
 
     @final
-    def get_cov(self) -> npt.NDArray[np.float64]:
-        """Gets the current covariance matrix."""
+    def get_cov(self, statistic: Optional[Statistic] = None) -> npt.NDArray[np.float64]:
```

Comment:
The application I wrote this for was to get error bars when plotting the data vector. The idea was specifically to abstract away the indices and instead use the statistics, since that's what the user interacts with. I see the use of passing a list of statistics though, to get their corresponding sub-matrix.

Reply:
We will expand the interface to have a single method `get_cov` that can accept:

1. a single `Statistic`
2. a list of `Statistic`
3. a single `np.ndarray` (the indices)
4. a list of `np.ndarray` (a list of indices)

In the case when a stat or list of stats is passed in, we need also to make sure that the stat (or all the stats) that have been  passed are in the likelihood object on which we've called the `get_cov` method. We will make the code verify this.

We also have to specify the order in which entries appear in the returned matrix. We propose to respect the order of the entries of the list of statistics (or of numpy arrays), so that the user-specified order of the list controls the ordering of the elements in the returned matrix result, rather than the order of the entires in the SACC data object controlling the ordering of the entries in the returned matrix.

For example, if we pass a list of stats of length 3 (note the order of the entries in this passed to `get_cov`:
stats1 -> 0:9, stats2 -> 10:19, stats3 -> 20:29 => `get_cov([stats1,stats3,stats2]) `-> 0:9 + 20:29 + 10:19